content/per-pool-locking.md

```diff
@@ -0,0 +1,143 @@
+on every single pool and this means that the time that it would take to list all of
+the pools or filesystems would increase proportionally with the number of pools
+on the system. Because locking is relatively expensive, we noticed a significant
+slowdown when listing larger numbers of pools and filesystems.
+
+Our ideal scenario was to have the benefits of a read-write lock so that list
+operations could run in parallel but to provide an ability to either lock single
+pools or all pools in one operation so that locking all pools would take the
+same amount of time no matter how many were present on the system.
+
+Design
+======
+After determining that no locking data structure like this appeared to exist
+in tokio, we took some time to look into how tokio implements its locking data
+structures. The API for much of the locking data structures appeared to
+be a lock acquisition method that returned a future. This future would poll the state
+of the lock and either update the internal data structures to indicate that
+the lock had been acquired or put itself to sleep until it was ready to poll
+again. The `drop` method on the data structure returned by the future would
+trigger waking up a task to poll again. This seemed perfectly workable with
+a more granular read-write lock. The only difference would be that we would
+need to keep track of locks on individual pools as well as locks on the entire
+collection. The proper locking conflict rules would need to be checked:
+
+* `WriteAll` conflicts with all other operations.
+* `ReadAll` conflicts with `WriteAll` and `Write` on any pool.
+* `Write` conflicts with `WriteAll` and `Read` or `Write` on the same pool.
+* `Read` conflicts with `WriteAll` and `Write` on the same pool.
+
+Any attempt to acquire two conflicting locks would queue one of the tasks to
+be woken up once the conflicting lock was dropped.
+
+Notable design choices
+======================
+We chose to implement our lock as a starvation-free lock. Implementing a lock
+that allows `ReadAll` to bypass `Write*` requests that are queued when another
+`ReadAll` request has already acquired the lock leads to behavior where `Write*`
+requests could block indefinitely. This behavior is particularly undesirable as
+this behavior could cause list operations to block filesystem extension handling
+indefinitely which could lead to IO errors and a full filesystem. A starvation-free
+locking approach puts a task in a FIFO queue if any are already queued in front of
+it. The notable downside of this is slightly more latency for handling locking
+requests, but the benefits seemed to outweigh this.
+
+Because tokio can cause spurious wake ups for tasks, we assign a unique integer
+ID to each future responsible for polling the lock for readiness. In the case
+where there is both a legitimate and spurious wake up at the same time, this
+allows our lock to differentiate between the two woken tasks to determine which
```

Comment:
The explanation for this gets highly technical and it might be too technical for this blog post. Would something that's still a little vague but gives more information like "the lock keeps track of the IDs of which tasks it has woken which allows it to differentiate between valid and spurious wakeups" be better? Also interested in your feedback here @drckeefe.

Reply:
The reader of this content should be familiar with the topics covered in this blog to gain benefit from it.  If they are familiar with the topics, I believe that they will ask the same question, as mulkieran did.    I had the same question while reading this.

Reply:
Does my suggestion clarify that for you at all @drckeefe?

Reply:
It does, if it means, each task is assigned a unique integer ID and the lock keeps track of IDs of which task it has woken.

Reply:
Yep! That's what it means.